### ISSUE TITLE
Promise AbstractStorageProvider

### ISSUE THREAD

**Issue by jamesikanos** [damienbod/angular-auth-oidc-client#841]
I'm looking at creating a PWA so I'm using Refresh Tokens. I, therefore, desire to move away from the default SessionStorage implementation.

IndexedDB and the CacheStorage API (the current recommendations for storing data within PWA's) are only available as asynchronous promise-based APIs.

I would like the AbstractStorageProvider to implement promise-based methods for Read/Write. When being used with SessionStorage/LocalStorage these can simply be returned with empty promises, not affecting the existing implementations.

I've looked into use LocalStorage but this is no longer recommended.
See here: https://dev.to/rdegges/please-stop-using-local-storage-1i04

Thank you

**Comment by damienbod** [comment]
@jamesikanos  Thanks for reporting.

This is a good enhancement. We need this. @FabianGosebrink agree?

Greetings Damien

**Comment by Johnnyrook777** [comment]
This also allows using the cordova secure storage - https://github.com/pradeep1991singh/cordova-plugin-secure-key-store#readme

**Comment by FabianGosebrink** [comment]
We wanted to redo the storage support anyway so I think this might be a good enhancement.

**Comment by jeffreyabecker** [comment]
Do you have any design goals on this? My current work-around is more than a bit of a kludge and I'll take a swing at implementing it.

**Comment by philjones88** [comment]
Just echoing support. I wrote an async storage class using Capacitor (using Ionic so like OP, local storage for refresh tokens isn't great) and their Storage plugin (see https://capacitorjs.com/docs/apis/storage ).

Here's what I wrote. I'll just have to await v12 :)

```
import { Injectable } from '@angular/core';
import { AbstractSecurityStorage } from 'angular-auth-oidc-client';
import { Plugins } from '@capacitor/core';


@Injectable()
export class CapacitorStorage implements AbstractSecurityStorage {
  public async read(key: string): Promise<any> {
    const result = await Plugins.Storage.get({ key });
    return JSON.parse(result.value);
  }

  public async write(key: string, value: any): Promise<boolean> {
    await Plugins.Storage.set({ key, value: JSON.stringify(value) });
    return true;
  }

  public async remove(key: string): Promise<boolean> {
    await Plugins.Storage.remove({ key });
    return true;
  }
}```

**Comment by SSchwaiger** [comment]
Hello!

I'm trying to store the refresh token using "keytar" in an Electron app. keytar only offers an asynchronous interface so... is this feature part of V12 or is it deplyed?

**Comment by damienbod** [comment]
Hi @SSchwaiger this will not make it in V12. We want to release V12 soon, and don't have the capacity to support this. The session storage and the local storage are synchronous which is the 95% use case of the lib. So supporting both synchronous and asynchronous would require a large effort to implement. 

We do plan this feature but have no timeline.

Greetings Damien

**Comment by marklagendijk** [comment]
I stumbled upon this and want to react to the claim that LocalStorage is 'not secure'.

> I've looked into use LocalStorage but this is no longer recommended.
See here: https://dev.to/rdegges/please-stop-using-local-storage-1i04

This article claims that LocalStorage is not secure, because if you are vulnerable to an XSS attack, the attacker could read the contents of the storage.
This is true, however, that doesn't mean that LocalStorage should never be used for storing sessions. It means that you need to protect against XSS attacks, which you also should do if you don't store sensitive stuff in LocalStorage.
If you comply with the following rules, you won't be easily vulnerable to an XSS attack:
* Use a proper [Content-Security-Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) (without `unsafe-*` directive for the `script-src`).
* Don't [support Internet Explorer](https://caniuse.com/contentsecuritypolicy), or if you must, only IE  >= 10, and add the header under the name X-Content-Security-Policy
* Use Angular or an other SPA framework for developing the web app. With frameworks like Angular it is hard to introduce an XSS vulnerability.
* Run the web app on its own (sub-)domain. This way only the web app can access the sensitive data, and the web app is protected by 2 very strong layers of security.

**Comment by trytuna** [comment]
+1 for this feature

**Comment by evancjohnson** [comment]
@FabianGosebrink, I hope all is well!

Thank you for contributing as much as you do to this library. I understand this a breaking change, I was wondering if any progress has been made on this? Is this something that I could help complete? I am working on an Ionic application, and I would love to be able to persist the token to the native platform's storage.

I know this would be a large

**Comment by FabianGosebrink** [comment]
Is this issue still relevant? Ionic _only_ supports Promises?

**Comment by evancjohnson** [comment]
> Is this issue still relevant? Ionic _only_ supports Promises?

IMO, I think at least supporting an observable implementation make sense so custom storage developers can choose to go down the async route or observable.

**Comment by Spinnenzunge** [comment]
Yes still relevant. We would love to use the Module in our App but we are using an Async Storage solution

**Comment by mariusbolik** [comment]
Would be nice if we could use [Capacitor Preferences](https://github.com/ionic-team/capacitor-plugins/tree/main/preferences) or [Capacitior Secure Storage](https://github.com/martinkasa/capacitor-secure-storage-plugin) to store Access Tokes securely.

**Comment by Spinnenzunge** [comment]
@FabianGosebrink I think I have a lightweight proposal that could solve the requirement of a lot of people subscribing to this issue.

Expose a method that allows us to programatically make an authorization call with a refresh token. With that method, we could store a refresh token with a long TTL in our own async secure storage and then after being able to rehydrate that storage call the authorization method by passing this refresh token. The authorization call would then either be successful and create its new state managed by the library or fail (because config changed, refresh token expired, etc.) and we could just stay on the login screen and have the user login again.

The code for getting a new auth state with a refresh token already exists, so you would just need to expose it to the client. Maybe `forceRefreshSession` would already be the correct method and just add an optional refresh token parameter to it.

What do you think, could that be an option instead of the whole refactoring of the storage as is?
